impl/sql/mutationstorage/queue.go

```diff
@@ -29,29 +30,74 @@ import (
 	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
 )
 
+// AddShards creates and adds new shards for queue writing to a domain.
+func (m *Mutations) AddShards(ctx context.Context, domainID string, shardIDs ...int64) error {
+	glog.Infof("mutationstorage: AddShard(%v, %v)", domainID, shardIDs)
+	for _, shardID := range shardIDs {
+		if _, err := m.db.ExecContext(ctx,
+			`INSERT INTO Shards (DomainID, ShardID, Enabled)  Values(?, ?, ?);`,
+			domainID, shardID, true); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // Send writes mutations to the leading edge (by sequence number) of the mutations table.
 func (m *Mutations) Send(ctx context.Context, domainID string, update *pb.EntryUpdate) error {
 	glog.Infof("mutationstorage: Send(%v, <mutation>)", domainID)
+	shardID, err := m.randShard(ctx, domainID)
+	if err != nil {
+		return err
+	}
 	mData, err := proto.Marshal(update)
 	if err != nil {
 		return err
 	}
 	// TODO(gbelvin): Implement retry with backoff for retryable errors if
 	// we get timestamp contention.
-	return m.send(ctx, domainID, mData, time.Now())
+	return m.send(ctx, domainID, shardID, mData, time.Now())
+}
+
+// randShard returns a random, enabled shard for domainID.
+func (m *Mutations) randShard(ctx context.Context, domainID string) (int64, error) {
+	// TODO(gbelvin): Cache these results.
+	var shardIDs []int64
+	rows, err := m.db.QueryContext(ctx,
+		`SELECT ShardID from Shards WHERE DomainID = ? AND Enabled = ?;`,
+		domainID, true)
+	if err != nil {
+		return 0, err
+	}
+	for rows.Next() {
+		var shardID int64
+		if err := rows.Scan(&shardID); err != nil {
+			return 0, err
+		}
+		shardIDs = append(shardIDs, shardID)
+	}
+	if err := rows.Err(); err != nil {
+		return 0, err
+	}
+	if len(shardIDs) == 0 {
+		return 0, status.Errorf(codes.NotFound, "No shard found for domain %v", domainID)
+	}
+
+	// Return a random shard.
+	return shardIDs[rand.Intn(len(shardIDs))], nil
 }
 
 // ts must be greater than all other timestamps currently recorded for domainID.
-func (m *Mutations) send(ctx context.Context, domainID string, mData []byte, ts time.Time) error {
+func (m *Mutations) send(ctx context.Context, domainID string, shardID int64, mData []byte, ts time.Time) error {
 	tx, err := m.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
 	if err != nil {
 		return err
 	}
 
 	var maxTime int64
 	if err := tx.QueryRowContext(ctx,
-		`SELECT COALESCE(MAX(Time), 0) FROM Queue WHERE DomainID = ?;`,
-		domainID).Scan(&maxTime); err != nil {
+		`SELECT COALESCE(MAX(Time), 0) FROM Queue WHERE DomainID = ? AND ShardID = ?;`,
```

Comment:
Can `Time` be nil?

Reply:
Time is a `NOT NULL` column, but `Max(Time)` is `Nil` if there are no rows.